### PR TITLE
add ability to block ip addresses that are submitting spam #1938

### PR DESCRIPTION
**Pull request by Sarthak5598**
#1938

### CODE REVIEW

**Review by DonnieBLT — CHANGES_REQUESTED** [review]
looking good, a few more tweaks please

File: website/models.py [hunk 1]
```diff
@@ -788,6 +788,18 @@ def __str__(self):
         return f"{self.user.username} - {self.amount} BACON"
 
 
+class BlockedIP(models.Model):
+    address = models.GenericIPAddressField(null=True, blank=True)
+    reason_for_block = models.TextField(blank=True, null=True, max_length=255)
+    address_range_start = models.GenericIPAddressField(null=True, blank=True)
+    address_range_end = models.GenericIPAddressField(null=True, blank=True)
+    user_agent_string = models.CharField(max_length=255, default="", null=True, blank=True)
+    count = models.IntegerField(default=1)
```
[the remark this answers is not in the document]
Comment (Sarthak5598) [reply website/models.py:797]: its for the number of requests a user made

File: blt/middleware/ip_restrict.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    def blocked_ip_ranges(self):
+        blocked_ip_ranges = cache.get("blocked_ip_ranges")
+        if blocked_ip_ranges is None:
+            blocked_ip_start = BlockedIP.objects.values_list("address_range_start", flat=True)
+            blocked_ip_end = BlockedIP.objects.values_list("address_range_end", flat=True)
+            blocked_ip_ranges = list(zip(blocked_ip_start, blocked_ip_end))
+            cache.set("blocked_ip_ranges", blocked_ip_ranges, timeout=86400)
+        return blocked_ip_ranges
+
+    def ip_in_range(self, ip, ip_ranges):
```
[the remark this answers is not in the document]
Comment (Sarthak5598) [reply blt/middleware/ip_restrict.py:35]: okay!

**Review by DonnieBLT — CHANGES_REQUESTED** [review]
Can you please update the migration? I think it's using the old name.

**Review by DonnieBLT — CHANGES_REQUESTED** [review]
One change

**Review by DonnieBLT — APPROVED** [review]

**Pull request merged**